Interactive cli

juampynr opened the pull request:
For #82 

This is raw and lacks error handling and best practices but already does something:

1. Upon compiling the Go application, you get a `drupal9ci` executable.
2. Copying this executable into a Drupal project and running it shows the following prompt:
```
/Users/jrequena/projects/drupal [master] ./drupal9ci 
Use the arrow keys to navigate: ↓ ↑ → ← 
? Select CI provider: 
  ▸ Bitbucket
    CircleCI
    GitHub Actions
    GitLab CI
    Travis CI
```
3. Selecting an option above performs the curl request to the installer command and executes it.

As I am writing this I think that I can simply run the script without having to download it since it is packaged within the executable. I will fix this in this pull request.

Once we have this in place we could:

* Set up a job that creates releases automatically out of tags. Users would simply download the executable and run it.
* Start replacing the bash scripts by actual Go code that we can test easily.

juampynr commented on cmd/root.go:
```diff
@@ -0,0 +1,111 @@
+import (
+	"fmt"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/cobra"
```
cobra is a tool to create CLI interfaces in Go.

juampynr commented on cmd/root.go:
```diff
@@ -0,0 +1,111 @@
+	"fmt"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
```
viper is a tool to manage configuration. It pairs well with cobra. I have not used it yet but decided to get it in already since I am sure that we will make use of it.

juampynr commented on cmd/root.go:
```diff
@@ -0,0 +1,120 @@
+	var setupScript string
+	var err error
+
+	if len(os.Args) > 1 {
```
Had to add this in order to set a CI provider at the CircleCI job that tests the application.

juampynr commented:
@fjgarlin thoughts about this so far?

fjgarlin commented:
Code-wise, it all looks good (my knowledge of `go` is limited but it's all very readable). 

I'd also expect as part of this PR changes in the readme file explaining how to use the new tool.

Looking at the code, do we need to compile the command with `go` first locally, and then run it?

fjgarlin commented:
1. You just addressed it
2. `Users would simply download the executable and run it.` that would address it.

I'll test it all locally, just wanted to give some quick feedback as I found it.

juampynr commented:
> Code-wise, it all looks good (my knowledge of `go` is limited but it's all very readable).
> 
> 1. I'd also expect as part of this PR changes in the readme file explaining how to use the new tool.
> 2. Looking at the code, do we need to compile the command with `go` first locally, and then run it?
> 
> -- _starts reading the very first comment..._

1: yep! I just adjusted that
2: no, I added a GitHub action that should create a release with the resulted binary that should be ready to use.

fjgarlin commented:
This is so awesome, it opens the door to future improvements like rewriting bash scripts in go, choosing the PHP version (once we figure out images), etc. And most importantly, it lowers the entry barrier to this project.

A super small suggestion. When you run the command, after choosing the platform, it takes around 5~10 seconds until you get the first line of output. For a second I wasn't sure if my machine was hanging, so it might be good to put a `Starting...` line so the user knows they'll need to wait a few seconds.

So something like:
```
? Select CI provider: 
  ▸ Bitbucket
    CircleCI
    GitHub Actions
    GitLab CI
    Travis CI
    
 ✔ Bitbucket
 Starting... 
 (5~10 seconds later)
 output: + wget -O...
```

Could be `Starting...` or `This might take a few seconds...` or something like that.

fjgarlin commented on cmd/root.go:
```diff
@@ -81,6 +57,24 @@ func Execute(setupScripts *scripts.SetupScripts) {
+	if len(args) > 1 {
+		return &args[1], nil
+	}
+	ciProviders := []string{"Bitbucket", "CircleCI", "GitHub Actions", "GitLab CI", "Travis CI"}
```
I'm new to `go` language. Can we not import/use the `const` defined in `scripts/scripts.go` to avoid repetition? If we can't that's totally fine.

fjgarlin commented:
You missed replacing the old command with the new one here: https://github.com/Lullabot/drupal9ci/blob/82-interactive-cli/README.md#github-actions

juampynr commented:
Will continue later.

fjgarlin commented:
Great work so far!! 💪

juampynr commented:
> This is so awesome, it opens the door to future improvements like rewriting bash scripts in go, choosing the PHP version (once we figure out images), etc. And most importantly, it lowers the entry barrier to this project.

You did read my mind! That's what we could do once this pull request lands.

juampynr commented:
@fjgarlin this is ready for another look.

juampynr commented:
@fjgarlin I will merge this tomorrow. Let me know if you find any issues. We can work on further changes in new pull requests.

fjgarlin commented:
I was away from the computer yesterday and couldn't check the activity. I've been checking the code changes and everything looks really solid. 

I think it's good to be merged. Approving the PR. Great work!

fjgarlin commented on .github/workflows/publish-docker-image.yml:
```diff
@@ -0,0 +1,21 @@
+
+jobs:
+  publish-docker-image:
+    name: Publish Docker image
```
Additional questions, but PR is good to be merged:
* Does this publish the image generated from `Dockerfile` into github? or is it another unrelated thing?
* If it's the image from `Dockerfile`, how would we use it later in the different platforms?

fjgarlin replied:
I think giving the user the option to choose might be enough. We can just give them options like this:
```
* Dockerfile9: php8.1, composer 2, npm 16...
* Dockerfile: php7.4, composer 1, npm 12...
* Choose your own image: this might not work out of the box. You'll be prompted to type it.
```

Then with that info, we can search/replace the config file of the platform they've used.

juampynr replied:
Yep! I like that.

fjgarlin commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,41 @@
+name: Test
```
Loving the additional tests

juampynr replied:
Yep! This is just the beginning. I want to move the shell scripts into easily testable Go code as you suggested.

Notice that I kind of hacked the default Cobra command: I removed sub-commands so you can just run `drupal9ci` but you can certainly have multiple sub commands so you could do, for example `drupal9ci check` to check for requirements and, when ready, run `drupal9ci install` to run the installer. We could also do this in a follow up.

juampynr replied:
At some point (sorry I am throwing out a lot of thoughts here), I would like to rename (again, I know) this repository to `drupalci` since we will be able to support multiple Drupal versions. Thoughts about this?

fjgarlin replied:
100% on this.

fjgarlin commented:
Not a problem. As this is not a package that is updated on people's projects, but rather a "use once, then tweak in your own copy", we will not be breaking people's projects. 

If new users find issues I'm sure they'll report them. But yeah, in any case, it's a good idea to give it a quick try by yourself and if I find time next week I can also do the same, but all of it as follow-ups.
